extensions/jta/runtime/src/main/java/org/apache/camel/quarkus/component/jta/TransactionalJtaTransactionPolicy.java

```diff
@@ -0,0 +1,105 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.quarkus.component.jta;
+
+import javax.inject.Inject;
+import javax.transaction.HeuristicMixedException;
+import javax.transaction.HeuristicRollbackException;
+import javax.transaction.RollbackException;
+import javax.transaction.Status;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+
+import org.apache.camel.CamelException;
+import org.apache.camel.jta.JtaTransactionPolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Helper methods for transaction handling
+ */
+public abstract class TransactionalJtaTransactionPolicy extends JtaTransactionPolicy {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TransactionalJtaTransactionPolicy.class);
+
+    @Inject
+    TransactionManager transactionManager;
+
+    protected void runWithTransaction(final Runnable runnable, final boolean isNew) throws Throwable {
+        if (isNew) {
+            begin();
+        }
+        try {
+            runnable.run();
+        } catch (Throwable e) {
+            rollback(isNew);
+            throw e;
+        }
+        if (isNew) {
+            commit();
+        }
+    }
+
+    private void begin() throws Exception {
+        transactionManager.begin();
+    }
+
+    private void commit() throws Exception {
+        try {
+            transactionManager.commit();
+        } catch (HeuristicMixedException | HeuristicRollbackException | RollbackException | SystemException e) {
+            throw new CamelException("Unable to commit transaction", e);
+        } catch (Exception | Error e) {
+            rollback(true);
+            throw e;
+        }
+    }
+
+    protected void rollback(boolean isNew) throws Exception {
+        try {
+            if (isNew) {
+                transactionManager.rollback();
+            } else {
+                transactionManager.setRollbackOnly();
+            }
+        } catch (Throwable e) {
+            LOG.warn("Could not rollback transaction!", e);
+        }
+    }
+
+    protected Transaction suspendTransaction() throws Exception {
+        return transactionManager.suspend();
+    }
+
+    protected void resumeTransaction(final Transaction suspendedTransaction) {
+        if (suspendedTransaction == null) {
+            return;
+        }
+
+        try {
+            transactionManager.resume(suspendedTransaction);
+        } catch (Throwable e) {
+            LOG.warn("Could not resume transaction!", e);
+        }
+    }
+
+    protected boolean hasActiveTransaction() throws Exception {
+        return transactionManager.getStatus() != Status.STATUS_MARKED_ROLLBACK
+                && transactionManager.getStatus() != Status.STATUS_NO_TRANSACTION;
+    }
```

Comment:
I guess we do not want end users to override these methods or even call them directly? Hence we should perhaps make them package visible and `final`? Maybe the subclasses of `TransactionalJtaTransactionPolicy` could be made `final` too?

Reply:
hmm, it makes sense.